### PR TITLE
os/bluestore: reduce memory footprint.

### PR DESCRIPTION
**Pull request by ifed01**
Primarily this includes following modifications:
1) Remove SharedBlob instance from non-shared blobs. Which also eliminates independent BufferSpace instance per each blob.
Instead non-shared blobs belonging to some Onode reference the same per-Onode BufferSpace.
Which required BufferSpace's key modification - now Buffer(s) are addressed via physical offsets not relative offsets within blobs.
2) Introduce 'squeezed' alternative implementation to bluestore_blob_t::extents field (PExtentVector). This field requires 40 bytes for empty value + additional 16 bytes per each pextent entry. Squeezing pextent entry to 8 bytes (by cutting off 12 low order bits for both offset and length fields (as we expect alloc size to be >= 4K) and using next 12 bits for length field only (resulting in 16MB for max blob size which seems to be enough) and reusing (via union) PExtentVector object space for fixed size array allows to keep up to 5 pextents within 40 bytes chunk. Given that per-blob pextent count is likely to be short (while blob count is significant) this gives some memory saving most of the time. Rollback to original PEXtentVector mechanics is supported as well.

Relevant modifications which were caused/inspired by the above ones:
1) read procedure refactoring
2) simplifying 'written' blob/buffers tracking.

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ifed01** [comment]
Below are some numbers for new memory usage:
These are mempool stats / cache hint counts after 300 secs of random access to 64K x 4MB prefilled objects via fio bluestore plugin.
1) Write only, default osd-memory target (4MB) (difference for bluestore_cache_other_XXX counters is the most notable)
Original:
            "bluestore_cache_onode_bytes": 45260800,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1585668926,
            "bluestore_cache_other_items": 29642198,
 
            "bluestore_onode_hits": 9455544,
            "bluestore_onode_misses": 77536,
            "bluestore_onode_shard_hits": 4577983,
            "bluestore_onode_shard_misses": 207244,

after the patch:
         "bluestore_cache_onode_bytes": 45793280,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1042679966,
            "bluestore_cache_other_items": 17693447,

       "bluestore_onode_hits": 9417668,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 4576834,
            "bluestore_onode_shard_misses": 180193,

2) Read-Write, default osd-memory target (4MB)
Original:
             "bluestore_cache_onode_bytes": 43324840,
            "bluestore_cache_onode_items": 63713,
            "bluestore_cache_other_bytes": 1215284846,
            "bluestore_cache_other_items": 23553493

           "bluestore_onode_hits": 3886638,
            "bluestore_onode_misses": 707760,
            "bluestore_onode_shard_hits": 2255965,
            "bluestore_onode_shard_misses": 1491410,

after the patch:
           "bluestore_cache_onode_bytes": 45793280,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1116098586,
            "bluestore_cache_other_items": 19050220,

           "bluestore_onode_hits": 4632884,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 3292640,
            "bluestore_onode_shard_misses": 591935,

3) Read-only, default osd-memory target (4MB)
Original:
        "bluestore_cache_onode_bytes": 44564480,
            "bluestore_cache_onode_items": 65536,
            "bluestore_cache_other_bytes": 1479594798,
            "bluestore_cache_other_items": 27844721,

         "bluestore_onode_hits": 6442706,
            "bluestore_onode_misses": 200196,
            "bluestore_onode_shard_hits": 5742173,
            "bluestore_onode_shard_misses": 1584324,

after the patch:
         "bluestore_cache_onode_bytes": 45088768,
            "bluestore_cache_onode_items": 65536,
            "bluestore_cache_other_bytes": 1132647161,
            "bluestore_cache_other_items": 19195776,

            "bluestore_onode_hits": 6893168,
            "bluestore_onode_misses": 65536,
            "bluestore_onode_shard_hits": 7015883,
            "bluestore_onode_shard_misses": 674282,

**Comment by ifed01** [comment]
The same as above but osd-memory target set to 8GB
1) Write only
Original:
         "bluestore_cache_onode_bytes": 45260800,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1688291007,
            "bluestore_cache_other_items": 31421052,
 
            "bluestore_onode_hits": 9437468,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 4588395,
            "bluestore_onode_shard_misses": 180194,

after the patch:
          "bluestore_cache_onode_bytes": 45793280,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1045187416,
            "bluestore_cache_other_items": 17747514,

            "bluestore_onode_hits": 9508708,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 4630089,
            "bluestore_onode_shard_misses": 180194,

2) Read-Write
Original:
            "bluestore_cache_onode_bytes": 45260800,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1776032033,
            "bluestore_cache_other_items": 33562048,

            "bluestore_onode_hits": 4687349,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 3335123,
            "bluestore_onode_shard_misses": 592334,

after the patch:
           "bluestore_cache_onode_bytes": 45793280,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1124069823,
            "bluestore_cache_other_items": 19152406,

            "bluestore_onode_hits": 4647942,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 3306339,
            "bluestore_onode_shard_misses": 594273,

3) Read-only
Original:
            "bluestore_cache_onode_bytes": 44564480,
            "bluestore_cache_onode_items": 65536,
            "bluestore_cache_other_bytes": 1786024574,
            "bluestore_cache_other_items": 33186371,

           "bluestore_onode_hits": 7121282,
            "bluestore_onode_misses": 65536,
            "bluestore_onode_shard_hits": 7272854,
            "bluestore_onode_shard_misses": 674516,

after the patch:
            "bluestore_cache_onode_bytes": 45088768,
            "bluestore_cache_onode_items": 65536,
            "bluestore_cache_other_bytes": 1143248613,
            "bluestore_cache_other_items": 19329392,
            "bluestore_fsck_bytes": 0,

            "bluestore_onode_hits": 6762230,
            "bluestore_onode_misses": 65536,
            "bluestore_onode_shard_hits": 6868987,
            "bluestore_onode_shard_misses": 677098,

**Comment by ifed01** [comment]
The same as above but osd-memory target set to 2GB
1) Write only
Original:
          "bluestore_cache_onode_bytes": 29240000,
            "bluestore_cache_onode_items": 43000,
            "bluestore_cache_other_bytes": 441301674,
            "bluestore_cache_other_items": 9210153,

           "bluestore_onode_hits": 5601078,
            "bluestore_onode_misses": 3177198,
            "bluestore_onode_shard_hits": 1051507,
            "bluestore_onode_shard_misses": 3320844,

after the patch:

           "bluestore_cache_onode_bytes": 38229408,
            "bluestore_cache_onode_items": 55566,
            "bluestore_cache_other_bytes": 397171310,
            "bluestore_cache_other_items": 8055111,

            "bluestore_onode_hits": 6721517,
            "bluestore_onode_misses": 2505795,
            "bluestore_onode_shard_hits": 1668668,
            "bluestore_onode_shard_misses": 2940690,

2) Read-Write
Original:
          "bluestore_cache_onode_bytes": 38553280,
            "bluestore_cache_onode_items": 56696,
            "bluestore_cache_other_bytes": 431413547,
            "bluestore_cache_other_items": 9473457,

            "bluestore_onode_hits": 6721517,
            "bluestore_onode_misses": 2505795,
            "bluestore_onode_shard_hits": 1668668,
            "bluestore_onode_shard_misses": 2940690,

after the patch:
           "bluestore_cache_onode_bytes": 45793280,
            "bluestore_cache_onode_items": 66560,
            "bluestore_cache_other_bytes": 1124069823,
            "bluestore_cache_other_items": 19152406,

            "bluestore_onode_hits": 4647942,
            "bluestore_onode_misses": 66560,
            "bluestore_onode_shard_hits": 3306339,
            "bluestore_onode_shard_misses": 594273,

3) Read-only
Original:
            "bluestore_cache_onode_bytes": 38734160,
            "bluestore_cache_onode_items": 56962,
            "bluestore_cache_other_bytes": 431241331,
            "bluestore_cache_other_items": 9408559,

           "bluestore_onode_hits": 1211475,
            "bluestore_onode_misses": 5047513,
            "bluestore_onode_shard_hits": 928083,
            "bluestore_onode_shard_misses": 5919265,

after the patch:
            "bluestore_cache_onode_bytes": 45088768,
            "bluestore_cache_onode_items": 65536,
            "bluestore_cache_other_bytes": 1143248613,
            "bluestore_cache_other_items": 19329392,

            "bluestore_onode_hits": 6762230,
            "bluestore_onode_misses": 65536,
            "bluestore_onode_shard_hits": 6868987,
            "bluestore_onode_shard_misses": 677098,

**Comment by markhpc** [comment]
@ifed01 This is great!  Your 4GB and 8GB tests showcase the reduction in memory usage while your 2GB tests actually show the gain in cached onodes.  That's wonderful!

Focusing on writes:

4GB before:
"bluestore_cache_other_bytes": 1585668926,
"bluestore_cache_other_items": 29642198,
avg bytes/item: ~53.5

4GB after:
"bluestore_cache_other_bytes": 1042679966,
"bluestore_cache_other_items": 17693447,
avg bytes/item: ~58.9

8GB before:
"bluestore_cache_other_bytes": 1688291007,
"bluestore_cache_other_items": 31421052,
avg bytes/item: ~53.7

8GB after:
"bluestore_cache_other_bytes": 1132647161,
"bluestore_cache_other_items": 19195776,
avg bytes/item: ~59.0

Do you expect the drop in cached items and corresponding increase in the average size of cached items is due to the changes to SharedBlob?

**Comment by markhpc** [comment]
@ifed01 also, when we combine this with #27705 we're going to see a multiplicative reduction in memory usage of the cache.  It's going to be glorious! :+1:

**Comment by ifed01** [comment]
> Do you expect the drop in cached items and corresponding increase in the average size of cached items is due to the changes to SharedBlob?

Not exactly. Squeezing pextents might impact this as well. Originally each pextent required additional 16 bytes in cached bytes and +1 item. Now most of them cost nothing (as they fit into the same 40 bytes that empty(!) PExtentVector object takes) hence cache items drop and (since pextent is relatively small) average item size raises.
But SharedBlob changes might impact this as well but rather in a less straightforward manner

**Comment by markhpc** [comment]
@ifed01 I see, thank you for the explanation!  I wonder if we should further break down bluestore_cache_other so we can see exactly what's changing here.

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -514,15 +532,46 @@ class BlueStore : public ObjectStore,
...
   struct Blob {
     MEMPOOL_CLASS_HELPERS();
 
     std::atomic_int nref = {0};     ///< reference count
     int16_t id = -1;                ///< id, for spanning blobs only, >= 0
     int16_t last_encoded_id = -1;   ///< (ephemeral) used during encoding only
-    SharedBlobRef shared_blob;      ///< shared blob state (if any)
+
+    const BufferSpaceBase* buffer_space() const {
+      if (unlikely(is_shared())) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.h]: my understanding from the linux kernel is that using likely/unlikely can have a high penalty if you're wrong, so it should really only be for exceptional cases.  i think cloned blobs aren't so uncommon?

**Comment by ifed01** [comment]
And some more numbers from OnodeSizeTracking test case (ceph_objectstore_tool).
It provides mempool stats for single 4M object filled in different ways and finally read back.
Original:
onode(680/1) other(23273/306)
onode(680/1) other(33811/1276)
onode(680/1) other(222227/3324)

after the patch
onode(688/1) other(15169/179)
onode(688/1) other(26731/1213)
onode(688/1) other(198763/3261)

File: src/os/bluestore/BlueStore.h (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.h]: yeah, probably you're right. will remove.

**Comment by markhpc** [comment]
@ifed01 I ran a quick 4k randwrite test on Officinalis across a 512GB data set using an 8GB memory target.  This is enough memory to keep all onodes in cache either way.  Since you already collected individual cache data I opted to look at performance and memory impact when not memory starved:
4K LibRBD Random Writes IOPS
4K LibRBD Random Writes 99 9th Percentile Latency
4K LibRBD Random Writes IOPS CPU Usage
4K LibRBD Random Writes RSS Memory

**Comment by ifed01** [comment]
@markhpc - Cool, thanks for sharing!

**Comment by ifed01** [comment]
And tail latency seems to be not that minor, from ~140 to to ~130 ms, i.e. around 7-8%.

**Comment by ifed01** [comment]
And tail latency seems to be not that minor, from ~140 to to ~130 ms, i.e. around 7-8%.

**Comment by ifed01** [comment]
@markhpc - would you mind to do the same for reads?

**Comment by markhpc** [comment]
> And tail latency seems to be not that minor, from ~140 to to ~130 ms, i.e. around 7-8%.

That's true!  Though Adam mentioned that he doesn't totally trust fio's latency measurements so perhaps take it with a grain of salt.

**Comment by markhpc** [comment]
> @markhpc - would you mind to do the same for reads?

Sure, right now I'm looking at randwrites with #27705 applied as well under low memory situations, but I will look at random reads as well.

**Comment by markhpc** [comment]
@ifed01 here are 4K rand read results:

4K LibRBD Random Reads IOPS
4K LibRBD Random Reads 99 9th Percentile Latency
4K LibRBD Random Reads CPU Usage
4K LibRBD Random Reads RSS Memory

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -1417,36 +1418,142 @@ BlueStore::BufferCacheShard *BlueStore::BufferCacheShard::create(
...
+{
+  for (auto it = from.buffer_map.begin(); it != from.buffer_map.end(); ++it) {
+    it->second->space = this;
+    ldout(get_cache()->cct, 20) << __func__ << " bufmap " << *it->second << dendl;
+    buffer_map[it->first] = it->second;
+  }
+  from.buffer_map.clear();
+}
+
+void BlueStore::BufferSpace::claim(BlueStore::BufferSpace& from,
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc:1463]:
```
2020-05-14T13:57:31.301 INFO:tasks.ceph.osd.7.smithi106.stderr: ceph version 16.0.0-1359-gb82b688c49e (b82b688c49e1ace1d53d03c387a7b08147555f10) pacific (dev)
2020-05-14T13:57:31.301 INFO:tasks.ceph.osd.7.smithi106.stderr: 1: (()+0x153c0) [0x7f9ac96ce3c0]
2020-05-14T13:57:31.301 INFO:tasks.ceph.osd.7.smithi106.stderr: 2: (()+0xfef3c9) [0x5581a08003c9]
2020-05-14T13:57:31.301 INFO:tasks.ceph.osd.7.smithi106.stderr: 3: (bluestore_blob_t::for_each_extent(bool, std::function<int (bluestore_pextent_t const&)>) const+0x69) [0x5581a0890799]
2020-05-14T13:57:31.301 INFO:tasks.ceph.osd.7.smithi106.stderr: 4: (BlueStore::BufferSpace::claim(BlueStore::BufferSpace&, bluestore_blob_t const&)+0xbd) [0x5581a07b4ced]
2020-05-14T13:57:31.302 INFO:tasks.ceph.osd.7.smithi106.stderr: 5: (BlueStore::Blob::make_blob_shared(BlueStore::Collection*, unsigned long)+0xe8) [0x5581a07ff8a8]
2020-05-14T13:57:31.302 INFO:tasks.ceph.osd.7.smithi106.stderr: 6: (BlueStore::ExtentMap::dup(BlueStore*, BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned lon
g&, unsigned long&, unsigned long&)+0xae2) [0x5581a0809fc2]
2020-05-14T13:57:31.302 INFO:tasks.ceph.osd.7.smithi106.stderr: 7: (BlueStore::_do_clone_range(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned long, unsigned
 long, unsigned long)+0x110) [0x5581a083b7e0]
2020-05-14T13:57:31.302 INFO:tasks.ceph.osd.7.smithi106.stderr: 8: (BlueStore::_clone(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, boost::intrusive_ptr<BlueStore::Onode>&)+0x300) [0x5581a0850bf0]
2020-05-14T13:57:31.302 INFO:tasks.ceph.osd.7.smithi106.stderr: 9: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ceph::os::Transaction*)+0x163c) [0x5581a0853a6c]
2020-05-14T13:57:31.303 INFO:tasks.ceph.osd.7.smithi106.stderr: 10: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool:
:TPHandle*)+0x2ad) [0x5581a085509d]
2020-05-14T13:57:31.303 INFO:tasks.ceph.osd.7.smithi106.stderr: 11: (non-virtual thunk to PrimaryLogPG::queue_transactions(std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<OpRequest>)+0x58) [0x5581a048d2d8]
2020-05-14T13:57:31.303 INFO:tasks.ceph.osd.7.smithi106.stderr: 12: (ReplicatedBackend::do_repop(boost::intrusive_ptr<OpRequest>)+0xc1e) [0x5581a065372e]
2020-05-14T13:57:31.303 INFO:tasks.ceph.osd.7.smithi106.stderr: 13: (ReplicatedBackend::_handle_message(boost::intrusive_ptr<OpRequest>)+0x19e) [0x5581a06650ce]
2020-05-14T13:57:31.303 INFO:tasks.ceph.osd.7.smithi106.stderr: 14: (PGBackend::handle_message(boost::intrusive_ptr<OpRequest>)+0x5b) [0x5581a04ce1cb]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 15: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x539) [0x5581a045d359]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 16: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x1c0) [0x5581a02e5a70]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 17: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x66) [0x5581a055be86]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 18: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xe8d) [0x5581a02e6c2d]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 19: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x403) [0x5581a09c1c83]
2020-05-14T13:57:31.304 INFO:tasks.ceph.osd.7.smithi106.stderr: 20: (ShardedThreadPool::WorkThreadSharded::entry()+0x14) [0x5581a09c4a84]
2020-05-14T13:57:31.305 INFO:tasks.ceph.osd.7.smithi106.stderr: 21: (()+0x9609) [0x7f9ac96c2609]
2020-05-14T13:57:31.305 INFO:tasks.ceph.osd.7.smithi106.stderr: 22: (clone()+0x43) [0x7f9ac9230103]
2020-05-14T13:57:31.305 INFO:tasks.ceph.osd.7.smithi106.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```

/a/kchai-2020-05-14_13:25:26-rados-wip-kefu-testing-2020-05-05-1333-distro-basic-smithi/5054362

**Comment by aclamk** [comment]
jenkins make check

**Comment by xiexingguo** [comment]
@ifed01 Can you rebase again?

**Comment by markhpc** [comment]
@ifed01 This would be a really good PR to get into quincy if we can.  Memory usage has come up a fair amount recently.

**Comment by tchaikov** [comment]
it's a shame that this change gets closed. i am dropping my request-to-change. i had test failures back then. see https://github.com/ceph/ceph/pull/32697#discussion_r425261476 

but things might have changed.

**Pull request closed without merging**